### ISSUE TITLE
How did you learn to javaparser?

### ISSUE THREAD

**Issue by yuandongjian** [javaparser/javaparser#3537]
There are so few document that I haven't found anything useful for a week. It's hard to learn without document.
I looked at Javaparservisited.pdf, but that only gives me a glimpse of Javaparser.

Just wanted to complain about javaparser's documentation.

**Comment by jlerbsc** [comment]
The documentation makes it possible to understand the operating principles or the possibilities of the project. What do you want to do with JavaParser?

**Comment by MysterAitch** [comment]
Hi @crazy-pizza ,

Where do you feel the gaps in documentation are? Is there anything specific you would like some help with? :)

Here is a broad outline of the knowledge and processes involved with using JavaParser, where each step _roughly_ relies on knowledge from all previous steps:

0. Foundational pre-knowledge of the tree data structure and with syntax trees (e.g., abstract/concrete syntax trees)
1. Foundational pre-knowledge of using Java, and with using code libraries in your own Java code (e.g., using Maven/Gradle to import JavaParser into your project)
2. Creating/configuring a JavaParser object in Java
3. Using JavaParser to create an AST and put it into a Java variable
4. Finding / searching for nodes within the syntax tree
5. Doing read-only analysis/logic based on these AST nodes
6. Changing the AST (rearranging, inserting, deleting, modifying nodes)
7. Using the JavaParser Symbol Solver to locate the _declaration_ of a given variable/reference
8. Further, more advanced, JavaParser usage or use of JavaParser within extreme circumstances/with particular requirements


Could you give a bit of information about which step(s) along this process need a bit more documentation please?

I suspect we might have useful documentation available for each of these and we might be able to share it with you, but if you've not already found it yourself then we might need to improve signposting to it :+1:

**Comment by yuandongjian** [comment]
@MysterAitch @jlerbsc Thank you for reply.
I want to parse javadoc and generated interface documentation for an entire project, my project might be complex and need to parse source jar packages at the same time. But I can only find a document that parses one Java file，I don't know how to analyze the dependencies between classes.

For example, here is my source code
``` source code
/**
 * there is my controller
 */
@Controller
public class MyController extends BaseController {


}

```

BaseController in a-source.jar
``` A.jar
/**
 * there is my baseController
 */
public abstract class BaseController {

    /**
     * send message
     * @param message the message info
     */
    @RequestMapping("/send")
    public void send(Message message) {

    }
}

```
Message in b-source.jar
``` C.jar
/**
 *  the message info
 */
public class Message {

    /**
     * the message body
     */
     private String body;

    /**
     * the message properties
     */
     private Map<String, String> properties;
}

```
How do I find the detailed interface comments for MyController？

I believe javaparser can solve my problem, but I can't find the documentation, please help me by providing the relevant documentation.

**Comment by MysterAitch** [comment]
A couple of things: 

- "parse" has a specific meaning, where a _single_ "compilation unit" (usually a file) is provided and transformed into a syntax tree.

- if providing JAR files, we are moving away from _parsing_ Java source code and moving toward the topic of symbol resolution -- taking a simple string reference (e.g., the name of a class or variable) and tracking back to find the location where this _symbol_ is declared

This means that you need to setup both JavaParser and the JavaParser Symbol Solver.


A requirement for this is that JavaParser needs to be provided with details about the source code and "other" files where types can be defined (e.g., JAR files).

To do this means providing a `TypeSolver` to the symbol solver. If you want to provide multiple type solvers, there is a `CombinedTypeSolver` available.


A simple example showing a simple (full) project using the `ReflectionTypeSolver` can be found within the sample repository here:

https://github.com/javaparser/javasymbolsolver-maven-sample


To add a JAR file, use a `JarTypeSolver` (usually by adding it to the `CombinedTypeSolver`).

This code sample from the ebook shows how to setup the type solver:

https://github.com/javaparser/javaparser-visited/blob/master/src/main/java/org/javaparser/examples/chapter5/UsingCombinedTypeSolver.java

```java
package org.javaparser.examples.chapter5;

import com.github.javaparser.symbolsolver.model.resolution.TypeSolver;
import com.github.javaparser.symbolsolver.resolution.typesolvers.CombinedTypeSolver;
import com.github.javaparser.symbolsolver.resolution.typesolvers.JarTypeSolver;
import com.github.javaparser.symbolsolver.resolution.typesolvers.JavaParserTypeSolver;
import com.github.javaparser.symbolsolver.resolution.typesolvers.ReflectionTypeSolver;

import java.io.File;
import java.io.IOException;

public class UsingCombinedTypeSolver {

    public static void main(String[] args) throws IOException {
        TypeSolver myTypeSolver = new CombinedTypeSolver(
                new ReflectionTypeSolver(),
                JarTypeSolver.getJarTypeSolver("jars/library1.jar"),
                JarTypeSolver.getJarTypeSolver("jars/library2.jar"),
                JarTypeSolver.getJarTypeSolver("jars/library3.jar"),
                new JavaParserTypeSolver(new File("src/main/java")),
                new JavaParserTypeSolver(new File("generated_code"))
        );
        // using myTypeSolver
    }
}
```

This code sample from the ebook shows how to use the typesolver:

https://github.com/javaparser/javaparser-visited/blob/master/src/main/java/org/javaparser/examples/chapter5/ResolveMethodCalls.java

```java
package org.javaparser.examples.chapter5;

import com.github.javaparser.StaticJavaParser;
import com.github.javaparser.ast.CompilationUnit;
import com.github.javaparser.ast.expr.MethodCallExpr;
import com.github.javaparser.symbolsolver.JavaSymbolSolver;
import com.github.javaparser.symbolsolver.model.resolution.TypeSolver;
import com.github.javaparser.symbolsolver.resolution.typesolvers.ReflectionTypeSolver;

import java.io.File;

public class ResolveMethodCalls {

    private static final String FILE_PATH = "src/main/java/org/javaparser/examples/chapter5/A.java";

    public static void main(String[] args) throws Exception {
        TypeSolver typeSolver = new ReflectionTypeSolver();

        JavaSymbolSolver symbolSolver = new JavaSymbolSolver(typeSolver);
        StaticJavaParser
                .getConfiguration()
                .setSymbolResolver(symbolSolver);

        CompilationUnit cu = StaticJavaParser.parse(new File(FILE_PATH));

        cu.findAll(MethodCallExpr.class).forEach(mce ->
                System.out.println(mce.resolve().getQualifiedSignature()));
    }
}
```


Hope this helps!

**Comment by yuandongjian** [comment]
@MysterAitch Yeah, I know that I should use JavaSymbolSolver and CombinedTypeSolver, as mentioned in JavaparServisited.pdf, But I can't find the next documentation on how to use CompilationUnit, just some very simple examples like VoidVisitorAdapter or cu.findAll(xxx.class). I don't even know how to get the value of annotation like @RequestMapping(value="/getSomething"). 

I tried to get some clues from debug, Javaparser document, and other projects that use Javaparser, but I got nothing, which frustrated me, and I can only guess which function I should use to get the data I want.

So I was curious to know how others learned Javaparser and applied it to their projects...  :)

**Comment by daniel-huss** [comment]
@crazy-pizza Hello, I think you'll be successful if you just keep doing what you're doing, and reward yourself with small victories *before* the frustration kicks in :)

For example, you could achieve something like "print a list all nodes that are annotated" and you'd end up with something like:
```java
    void foo1(CompilationUnit cu) {
        for (var n : cu.findAll(Node.class, it -> it instanceof NodeWithAnnotations)) {
            System.out.println(n.getClass().getName());
        }
    }
```
The key discovery would be "ok, AST node capabilities are often expressed by implementing an interface called NodeWithXXX".

Then you could move on to check out the API of `NodeWithAnnotations` and find that it provides `getAnnotations()`. It returns some `AnnotationExpr` instances. Again, capabilities that go beyond providing the annotation name are in more specialized subtypes. So what kind of annotation expressions exist? 

![th](https://user-images.githubusercontent.com/56636112/162446637-37be70af-9f13-4ccd-8479-67756ad99d84.png)

Okay, so let's print out the annotation values next:

```java
    void foo3(CompilationUnit cu) {
        for (var n : cu.findAll(Node.class, it -> it instanceof NodeWithAnnotations)) {
            for (var annotationExpr : ((NodeWithAnnotations<?>) n).getAnnotations()) {
                if (annotationExpr instanceof SingleMemberAnnotationExpr) {
                    System.out.println("Single: " + ((SingleMemberAnnotationExpr) annotationExpr).getMemberValue());
                } else if (annotationExpr instanceof NormalAnnotationExpr) {
                    System.out.println("Pairs: " + ((NormalAnnotationExpr) annotationExpr).getPairs());
                }
            }
        }
    }
```

So, TL;DR: Small steps, I guess.

**Comment by jlerbsc** [comment]
There were several answers to this question which should allow to close the issue.
